README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ggshakeR <img src="man/figures/logo.png" align="right" height="150">

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![License:
MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![check-standard](https://github.com/abhiamishra/ggshakeR/workflows/R-CMD-check/badge.svg)](https://github.com/abhiamishra/ggshakeR/actions)
[![codecov](https://codecov.io/gh/abhiamishra/ggshakeR/branch/main/graph/badge.svg?token=GO5918U56P)](https://codecov.io/gh/abhiamishra/ggshakeR)
<!-- badges: end -->

Welcome to `ggshakeR`!

This is R’s all-inclusive package for visualization and analysis of
publicly available soccer data.

This package aims to provide people interested in football analysis and
visualization a platform to learn about it while also learning a new
language at the same time. This package is designed to work with free
soccer data and, for now, include:

-   [FBref](https://fbref.com/en/)
-   [StatsBomb Open Dataset](https://github.com/statsbomb/StatsBombR)
-   [Understat](https://understat.com/)

Let’s get you started with the package!

# Getting Started With ggshakeR

First things first, install R and RStudio. R is the actual language
while RStudio is the IDE that will help you work with the R programming
language in a nice way.

Here are links to download both:

-   [R](https://www.r-project.org/)
-   [RStudio](https://www.rstudio.com/products/rstudio/download/)

OK, now how do you get ggshakeR into RStudio?

First install the devtools package:

``` r
install.packages("devtools")
```

After that, write the actual code to install ggshakeR:

``` r
devtools::install_github("abhiamishra/ggshakeR")
```

There were many **BIG** changes made in **version 0.2.0**. Do note that
the changes are permanent and will be the standard going forward. So
please take the time to read the [Guide to Version 0.2.0 Vignette]() and
the package documentation carefully so that you can transition over to
the new argument syntax and make use of the new functionality that
version `0.2.0` (and beyond!) will provide. If you want to go back to
version `0.1.2` after you’ve already installed `0.2.0` (because you’re
not yet ready to edit your scripts to the new syntax or whatever), you
can **re-install** version `0.1.2` by running this:

``` r
devtools::install_github("abhiamishra/ggshakeR@0.1.2")
```

Now the package is installed, but it’s not in your session just yet!
When a package is in a session means you can actually use the package:

``` r
library(ggshakeR)
```

And that’s it you’re done!

We hope you have fun!

## Helpful Links

Here are some helpful links to get you started with the package:

-   [Guide To Pitch
    Plots](https://abhiamishra.github.io/ggshakeR/articles/Guide_to_Pitch_Plots.html)
-   [Guide To Expected
    Threat](https://abhiamishra.github.io/ggshakeR/articles/Guide_to_Exp_Threat.html)
-   [Guide To Pizza
    Plots](https://abhiamishra.github.io/ggshakeR/articles/Guide_to_PizzaPlots.html)
-   [Guide To Expected Possession
    Value](https://abhiamishra.github.io/ggshakeR/articles/Guide_to_EPV.html)
-   [Guide To Version
    0.2.0](https://abhiamishra.github.io/ggshakeR/articles/Guide_to_Version_0-2-0.html)

# Credit

Thank you to StatsBomb and [Ewan Henderson](https://github.com/ewenme)
for their respective packages. A big thank you to Ninad Barbadikar for
creating the beautiful logo and helping with the package in general.
Thank you [Jase Ziv](https://github.com/JaseZiv) for your help and your
package [WorldFootballR](https://github.com/JaseZiv/worldfootballR) from
which this took inspiration from.

# Contact and Support

Tell your friends about ggshakeR and shake theiR lives away! You can
contact me at [email](abhiamishra0@gmail.com) or at my
[Twitter](https://twitter.com/MishraAbhiA)

Powered by Chatalytics:tm:

Owner and Creator: [**Abhishek Amol
Mishra**](https://github.com/abhiamishra)

Authors: [**Harsh Krishna**](https://github.com/harshkrishna17) & [**Ryo
Nakagawara**](https://github.com/Ryo-N7)

# Licenses Used

-   [Torvaney](https://github.com/Torvaney/ggsoccer) and the ggsoccer
    package has been used in the package for creation of pitches. The
    use of this adheres to the [MIT
    License](https://github.com/Torvaney/ggsoccer/blob/master/LICENSE.md)
    provided.

Licensing Copyright Statement: Copyright (c) 2019 Benjamin Torvaney

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the
“Software”), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


NEWS.md
# ggshakeR 0.2.0.9002

- Bug Fixes
  - Fixes bugs in the `plot_pizza` in the custom option of the comparison plot [issue #123](https://github.com/abhiamishra/ggshakeR/issues/123). 

# ggshakeR 0.2.0.9001

- Bug Fixes
	- Fixed `plot_pizza` bugs caused by FBRef updates [issue #121](https://github.com/abhiamishra/ggshakeR/issues/121)
		- Changed stat selection to named stats instead of selecting row numbers, which change because of FBRef data changes.
		- Added method to de-duplicate some stats that were returned for multiple positions, causing errors.
		- Changed season default argument to "Last 365 Days Men's Big 5 Leagues, UCL, UEL" to match FBRef data for men's big 5 league players. (this makes the caption wider than before, so some have been split over two lines)
		- Replaced long `|`-statements with `%in%`.
		- Deleted some old unused code.

# ggshakeR 0.2.0.9000

-   Bug Fixes
    - Fixed `plot_heatmap` verification assessment [issue #101](https://github.com/abhiamishra/ggshakeR/issues/101)
    - Fixed pkgdown build issues [issue #104](https://github.com/abhiamishra/ggshakeR/issues/104)
    - Fixed `plot_passnet` direction orientation [issue #108](https://github.com/abhiamishra/ggshakeR/issues/108)
    - Fixed `plot_heatmap` and `plot_passflow`direction orientation + `plot_pass`opta functionality was fixed.[issue #111](https://github.com/abhiamishra/ggshakeR/issues/111)
    - Added "Created using ggshakeR" to `plot_heatmap`, `plot_passflow`, and `plot_pass`[issue #110](https://github.com/abhiamishra/ggshakeR/issues/110)
    - Added "Created using ggshakeR" to the rest of the functions [issue #110](https://github.com/abhiamishra/ggshakeR/issues/110)
    - Converted `plot_pizza` to work with new FBRef data from Opta. [issue #118](https://github.com/abhiamishra/ggshakeR/issues/118)

# ggshakeR 0.2.0

-   Standardized function argument and object names
    -   Use snake_case: `binSize` to `bin_size`, `dataType` to `data_type`, etc.
    -   Data arguments all conform to `data`
    -   Use US spelling: `color` rather than `colour`, etc.
    -   Argument names made longer for clarity: `subt_size` to `subtitle_size`, `roll_avg` to `rolling_average`, etc.
    -   See [issue #27](https://github.com/abhiamishra/ggshakeR/issues/27)
-   Created new functions
    -   `plot_convexhull()` & `hull_fun()` utility function: See [issue #28](https://github.com/abhiamishra/ggshakeR/issues/28)
    -   `plot_voronoi()` function: See [issue #28](https://github.com/abhiamishra/ggshakeR/issues/28)
    -   `plot_passnet()` function: See [issue #31](https://github.com/abhiamishra/ggshakeR/issues/31), [issue #53](https://github.com/abhiamishra/ggshakeR/issues/53)
    -   `calculate_epv()` function: See [issue #25](https://github.com/abhiamishra/ggshakeR/issues/25)
    -   Created vignettes for new functions: See [issue #45](https://github.com/abhiamishra/ggshakeR/issues/45)
-   Improved functions
    -   Improved `plot_pizza()`: See [issue #37](https://github.com/abhiamishra/ggshakeR/issues/37)
    -   Added `jdp` pitch type option for `plot_heatmap()`: See [issue #37](https://github.com/abhiamishra/ggshakeR/issues/37)
    -   Simplified `plot_pass()`: See [issue #64](https://github.com/abhiamishra/ggshakeR/issues/64)
-   Improved test coverage
    -   See [issue #44](https://github.com/abhiamishra/ggshakeR/issues/44)
-   Other minor fixes

R/plot_pizza.R
#' Function for plotting percentile/pizza plots
#'
#' This function allows for data, that has to be scraped from FBref, to be used
#' for plotting single and comparison percentile plots.
#'
#' For best image quality use:
#' ggsave("image.png", width = 2900, height = 2800, units = "px")
#'
#' @param data Data frame can contain either one player or two depending on the type of plot made
#' @param type Type of plot -> single and comparison
#' @param template Selecting a group of pre-selected metrics for each position by position namely:
#'                 outfielder, goalkeeper and custom
#' @param color_possession Selecting the color for possession group of stats. To be used only for single player plot
#' @param color_attack Selecting the color for attacking group of stats. To be used only for single player plot
#' @param color_defense Selecting the color for defense group of stats. To be used only for single player plot
#' @param player_1 Selecting the first player. To be used only for comparison plot
#' @param player_2 Selecting the second player. To be used only for comparison plot
#' @param color_compare Selecting the color of comparison to be used only for comparison plot
#' @param season Specify what season to pick for a single player pizza chart. Pick the scouting period from the scouting period column in the data
#' @param season_player_1 Specify what season to pick for the first player in a pizza chart
#' @param season_player_2 Specify what season to pick for the second player in a pizza chart
#' @param theme Specify the theme of the pizza chart -> dark, black, and white. Default set to dark
#' @return a ggplot2 object
#'
#' @import dplyr
#' @import ggplot2
#' @importFrom stringi stri_wrap stri_trans_general
#'
#' @export
#'
#' @examples
#' \dontrun{
#' plot1 <- plot_pizza(data = data, type = "comparison", template = "outfielder",
#'                     player_1 = "Nicolo Barella", player_2 = "Ilkay Gundogan",
#'                     season_player_1 = "Last 365 Days Men's Big 5 Leagues, UCL, UEL", 
#'                     season_player_2 = "Last 365 Days Men's Big 5 Leagues, UCL, UEL",
#'                     color_compare = "lightgreen", theme = "black")
#' plot1
#'
#' plot2 <- plot_pizza(data = data1, type = "single", template = "outfielder",
#'                     color_possession = "green", color_attack = "lightblue", 
#'                     season = "Last 365 Days Men's Big 5 Leagues, UCL, UEL",
#'                     color_defense = "red", theme = "dark")
#' plot2
#' }

plot_pizza <- function(data, type = "", template, 
                       color_possession = "#41ab5d", color_attack = "#2171b5", 
                       color_defense = "#fec44f", 
                       player_1, player_2, 
                       color_compare = "#41ab5d", 
                       season = "Last 365 Days Men's Big 5 Leagues, UCL, UEL", 
                       season_player_1 = "Last 365 Days Men's Big 5 Leagues, UCL, UEL", 
                       season_player_2 = "Last 365 Days Men's Big 5 Leagues, UCL, UEL", 
                       theme = "") {
  
  if (theme == "dark" || theme == "") {
    fill_b <- "#0d1117"
    color_b <- "#0d1117"
    colorText <- "white"
    gridline <- "565656"
    colorLine <- "white"
  } else if (theme == "black") {
    fill_b <- "black"
    color_b <- "black"
    colorText <- "white"
    gridline <- "565656"
    colorLine <- "white"
  } else if (theme == "white") {
    fill_b <- "white"
    color_b <- "white"
    colorText <- "black"
    gridline <- "565656"
    colorLine <- "black"
  }
  
  if (type == "single" || type == "") { ## SINGLE PLOT ----
    
    data <- data %>%
      filter(scouting_period == season)
    
    if (template == "outfielder") {
      
      data_selected <- data %>% 
        mutate(stat = case_when(
          Statistic == "Non-Penalty Goals" |
            Statistic == "Non-Penalty xG" |
            Statistic == "Shots Total" |
            Statistic == "Assists" |
            Statistic == "xAG" |
            Statistic == "npxG + xAG" ~ "Attacking",
          Statistic == "Shot-Creating Actions" |
            Statistic == "Passes Attempted" |
            Statistic == "Pass Completion %" |
            Statistic == "Progressive Passes" |
            Statistic == "Dribbles Completed" |
            Statistic == "Touches (Att Pen)" |
            Statistic == "Progressive Passes Rec" ~ "Possession",
          Statistic == "Tackles" |
            Statistic == "Interceptions" |
            Statistic == "Blocks" |
            Statistic == "Clearances" |
            Statistic == "Aerials won" ~ "Defending",
          TRUE ~ NA_character_
        )) %>%
        drop_na(stat) %>%
        distinct(Statistic, .keep_all = TRUE)
      
    }
    else if (template == "goalkeeper") {
      
      data_selected <- data %>%
        mutate(stat = case_when(Statistic == "Goals Against" |
                                  Statistic == "PSxG/SoT" |
                                  Statistic == "Save Percentage" |
                                  Statistic == "PSxG-GA" ~ "Defending",
                                Statistic == "Touches" |
                                  Statistic == "Launch %" |
                                  Statistic == "Goal Kicks" |
                                  Statistic == "Avg. Length of Goal Kicks" ~ "Possession",
                                Statistic == "Crosses Stopped %" |
                                  Statistic == "Def. Actions Outside Pen. Area" |
                                  Statistic == "Avg. Distance of Def. Actions" ~ "Attacking",
                                TRUE ~ NA_character_)) %>%
        drop_na(stat) %>%
        distinct(Statistic, .keep_all = TRUE)
      
    }
    else if (template == "custom") {
      
      data_selected <- data %>% 
        mutate(stat = case_when(
          StatGroup == "Standard" ~ "Attacking", 
          StatGroup == "Shooting" ~ "Attacking",
          StatGroup == "Passing" ~ "Possession",
          StatGroup == "Pass Types" ~ "Possession",
          StatGroup == "Goal and Shot Creation" ~ "Possession",
          StatGroup == "Defense" ~ "Defending",
          StatGroup == "Possession" ~ "Possession",
          StatGroup == "Miscellaneous Stats" ~ "Defending",
          TRUE ~ NA_character_
        ))
      
    }
    
    #    if (template == "forward") {
    #      
    #     if (nrow(data) > 148) {
    #        
    #        data_selected <- data[c(3, 8, 13, 24, 42, 128, 45, 115, 133, 107, 101, 102, 26, 147), ]
    #        
    #      } else {
    #        
    #        data_selected <- data[c(3, 8, 13, 24, 41, 127, 44, 114, 132, 106, 100, 101, 25, 146), ]
    #        
    #      }
    #    } else if (template == "midfielder") {
    #      
    #      if (nrow(data) > 148) {
    #        
    #        data_selected <- data[c(3, 9, 10, 13, 53, 44, 47, 116, 125, 133, 146, 147, 107, 96), ]
    #        
    #      } else {
    #        
    #        data_selected <- data[c(3, 9, 10, 13, 52, 43, 46, 115, 124, 132, 145, 146, 106, 97), ]
    #        
    #      }
    #    } else if (template == "defender") {
    #      
    #      if (nrow(data) > 148) {
    #        
    #        data_selected <- data[c(3, 11, 13, 44, 47, 129, 125, 110, 88, 96, 102, 106, 147, 108), ]
    #        
    #      } else {
    #        
    #        data_selected <- data[c(3, 11, 13, 43, 46, 128, 124, 109, 87, 95, 101, 105, 146, 107), ]
    #        
    #      }
    #    } else if (template == "full back") {
    #      
    #      if (nrow(data) > 148) {
    #        
    #        data_selected <- data[c(3, 9, 10, 13, 114, 46, 47, 125, 43, 44, 147, 96, 107, 102), ]
    #        
    #      } else {
    #        
    #        data_selected <- data[c(3, 9, 10, 13, 113, 45, 46, 124, 42, 43, 146, 95, 106, 101), ]
    #        
    #      }
    #    } else if (template == "winger") {
    #      
    #      if (nrow(data) > 148) {
    #        
    #        data_selected <- data[c(3, 22, 24, 42, 143, 45, 119, 47, 124, 133, 107, 146, 101, 102), ]
    #        
    #      } else {
    #        
    #        data_selected <- data[c(3, 21, 23, 41, 142, 44, 118, 46, 123, 132, 106, 145, 100, 101), ]
    #        
    #      }
    #    } else if (template == "goalkeeper") {
    #      
    #      if (nrow(data) > 36) {
    #        
    #        data_selected <- data[c(4, 20, 22, 24, 26, 29, 35, 36, 37), ]
    #        data_selected <- data_selected %>%
    #          mutate(stat = case_when(Statistic == "Save%" |
    #                                    Statistic == "PSxG" |
    #                                    Statistic == "PSxG-GA" ~ "Defending",
    #                                  Statistic == "Passes Attempted (Launched)" |
    #                                    Statistic == "Passes Attempted" |
    #                                    Statistic == "Average Pass Length" ~ "Possession",
    #                                  TRUE ~ "Attacking"))
    #      } else {
    #        
    #        data_selected <- data[c(4, 19, 21, 23, 25, 28, 34, 35, 36), ]
    #        data_selected <- data_selected %>%
    #          mutate(stat = case_when(Statistic == "Save%" |
    #                                    Statistic == "PSxG" |
    #                                    Statistic == "PSxG-GA" ~ "Defending",
    #                                  Statistic == "Passes Attempted (Launched)" |
    #                                    Statistic == "Passes Attempted" |
    #                                    Statistic == "Average Pass Length" ~ "Possession",
    #                                  TRUE ~ "Attacking"))
    #      }
    #    } else if (template == "custom") {
    #      data_selected <- data
    #    }
    
    player_name <- unique(data$Player)
    title <- paste(player_name, "Percentile Chart")
    min <- unique(data$BasedOnMinutes)
    sub <- unique(data$Versus)
    sub1 <- unique(data$scouting_period)
    subtitle <- paste("Compared to", sub, "|", sub1, "|", min, "minutes played")
    caption <- "Plot code by @RobinWilhelmus\nData from Stats Perform via FBref. Inspired by @NathanAClark. Created using ggshakeR."
    
    x <- c(data_selected$Statistic, data_selected$stat)
    
    data_selected <- data_selected %>%
      arrange(desc(stat), desc(Percentile)) %>%
      mutate(Statistic = factor(Statistic, levels = Statistic))
    
    ggplot(data_selected, aes(Statistic, Percentile)) +
      geom_bar(aes(y = 100, fill = stat), stat = "identity", width = 1, color = fill_b,
               alpha = 0.1, show.legend = FALSE) +
      geom_bar(stat = "identity", width = 1, aes(fill = stat), color = fill_b, alpha = 1) +
      coord_polar(clip = "off") +
      geom_hline(yintercept = 25, color = colorLine, linetype = "dashed", alpha = 0.8) +
      geom_hline(yintercept = 50, color = colorLine, linetype = "dashed", alpha = 0.8) +
      geom_hline(yintercept = 75, color = colorLine, linetype = "dashed", alpha = 0.8) +
      scale_fill_manual(values = c("Possession" = color_possession,
                                   "Attacking" = color_attack,
                                   "Defending" = color_defense)) +
      geom_label(aes(y = 90, label = Per90, fill = stat), size = 3, color = fill_b, show.legend = FALSE) +
      scale_y_continuous(limits = c(-20, 100)) +
      labs(fill = "",
           caption = caption,
           title = title,
           subtitle = subtitle) +
      theme_minimal() +
      theme(plot.background = element_rect(fill = fill_b, color = color_b),
            panel.background = element_rect(fill = fill_b, color = color_b),
            legend.position = "bottom",
            axis.title.y = element_blank(),
            axis.title.x = element_blank(),
            axis.text.y = element_blank(),
            axis.text.x = element_text(size = 12, color = colorText),
            text = element_text(color = colorText, size = 20),
            plot.title = element_text(hjust = 0.5, size = 26, color = colorText, face = "bold"),
            plot.subtitle = element_text(hjust = 0.5, size = 20, color = colorText),
            plot.caption = element_text(hjust = 0.5, size = 15, color = colorText),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank()) +
      scale_x_discrete(labels = function(x) text_wrap(x = x))
    
  } else if (type == "comparison") { ## COMPARISON PLOT ----
    
    data$Player <- stri_trans_general(str = data$Player, id = "Latin-ASCII")
    
    if (template == "outfielder") {
      
      data <- data %>% 
        mutate(stat = case_when(
          Statistic == "Non-Penalty Goals" |
            Statistic == "Non-Penalty xG" |
            Statistic == "Shots Total" |
            Statistic == "Assists" |
            Statistic == "xAG" |
            Statistic == "npxG + xAG" ~ "Attacking",
          Statistic == "Shot-Creating Actions" |
            Statistic == "Passes Attempted" |
            Statistic == "Pass Completion %" |
            Statistic == "Progressive Passes" |
            Statistic == "Dribbles Completed" |
            Statistic == "Touches (Att Pen)" |
            Statistic == "Progressive Passes Rec" ~ "Possession",
          Statistic == "Tackles" |
            Statistic == "Interceptions" |
            Statistic == "Blocks" |
            Statistic == "Clearances" |
            Statistic == "Aerials won" ~ "Defending",
          TRUE ~ NA_character_)) %>%
        drop_na(stat)
      
      data1 <- data %>%
        filter(Player == player_1) %>%
        filter(scouting_period == season_player_1) %>%
        distinct(Statistic, .keep_all = TRUE)
      data2 <- data %>%
        filter(Player == player_2) %>%
        filter(scouting_period == season_player_2) %>%
        distinct(Statistic, .keep_all = TRUE)
      
    }
    else if (template == "goalkeeper") {
      
      data <- data %>%
        mutate(stat = case_when(Statistic == "Goals Against" |
                                  Statistic == "PSxG/SoT" |
                                  Statistic == "Save Percentage" |
                                  Statistic == "PSxG-GA" ~ "Defending",
                                Statistic == "Touches" |
                                  Statistic == "Launch %" |
                                  Statistic == "Goal Kicks" |
                                  Statistic == "Avg. Length of Goal Kicks" ~ "Possession",
                                Statistic == "Crosses Stopped %" |
                                  Statistic == "Def. Actions Outside Pen. Area" |
                                  Statistic == "Avg. Distance of Def. Actions" ~ "Attacking",
                                TRUE ~ NA_character_)) %>%
        drop_na(stat)
      
      data1 <- data %>%
        filter(Player == player_1) %>%
        filter(scouting_period == season_player_1) %>%
        distinct(Statistic, .keep_all = TRUE)
      data2 <- data %>%
        filter(Player == player_2) %>%
        filter(scouting_period == season_player_2) %>%
        distinct(Statistic, .keep_all = TRUE)
      
    }
    else if (template == "custom") {
      
      data <- data %>% 
        mutate(stat = case_when(
          StatGroup == "Standard" ~ "Attacking", 
          StatGroup == "Shooting" ~ "Attacking",
          StatGroup == "Passing" ~ "Possession",
          StatGroup == "Pass Types" ~ "Possession",
          StatGroup == "Goal and Shot Creation" ~ "Possession",
          StatGroup == "Defense" ~ "Defending",
          StatGroup == "Possession" ~ "Possession",
          StatGroup == "Miscellaneous Stats" ~ "Defending",
          TRUE ~ NA_character_
        ))
      
      data1 <- data %>%
        filter(Player == player_1) %>%
        filter(scouting_period == season_player_1)
      data2 <- data %>%
        filter(Player == player_2) %>%
        filter(scouting_period == season_player_2)
      
    }
    
    #    if (template == "forward") {
    #      
    #      if (nrow(data1) > 148) {
    #        
    #        data1$no <- 1:nrow(data1)
    #        data2$no <- 1:nrow(data2)
    #        data1 <- data1[c(3, 8, 13, 24, 42, 128, 45, 115, 133, 107, 101, 102, 26, 147), ]
    #        data2 <- data2[c(3, 8, 13, 24, 42, 128, 45, 115, 133, 107, 101, 102, 26, 147), ]
    #      } else {
    #        
    #        data1$no <- 1:nrow(data1)
    #        data2$no <- 1:nrow(data2)
    #        data1 <- data1[c(3, 8, 13, 23, 41, 127, 44, 114, 132, 106, 100, 101, 25, 146), ]
    #        data2 <- data2[c(3, 8, 13, 24, 42, 128, 45, 115, 133, 107, 101, 102, 26, 147), ]
    #      }
    #    } else if (template == "midfielder") {
    #      
    #      if (nrow(data1) > 148) {
    #        
    #        data1$no <- 1:nrow(data1)
    #        data2$no <- 1:nrow(data2)
    #        data1 <- data1[c(3, 9, 10, 13, 53, 44, 47, 116, 125, 133, 146, 147, 107, 96), ]
    #        data2 <- data2[c(3, 9, 10, 13, 53, 44, 47, 116, 125, 133, 146, 147, 107, 96), ]
    #      } else {
    #        
    #        data1$no <- 1:nrow(data1)
    #        data2$no <- 1:nrow(data2)
    #        data1 <- data1[c(3, 9, 10, 13, 52, 43, 46, 115, 124, 132, 145, 146, 106, 95), ]
    #        data2 <- data2[c(3, 9, 10, 13, 53, 44, 47, 116, 125, 133, 146, 147, 107, 96), ]
    #      }
    #    } else if (template == "defender") {
    #      
    #      if (nrow(data1) > 148) {
    #        
    #        data1$no <- 1:nrow(data1)
    #        data2$no <- 1:nrow(data2)
    #        data1 <- data1[c(3, 11, 13, 44, 47, 129, 125, 110, 88, 96, 102, 106, 147, 108), ]
    #        data2 <- data2[c(3, 11, 13, 44, 47, 129, 125, 110, 88, 96, 102, 106, 147, 108), ]
    #        
    #      } else {
    #        
    #        data1$no <- 1:nrow(data1)
    #        data2$no <- 1:nrow(data2)
    #        data1 <- data1[c(3, 11, 13, 43, 46, 128, 124, 109, 87, 95, 101, 105, 146, 107), ]
    #        data2 <- data2[c(3, 11, 13, 44, 47, 129, 125, 110, 88, 96, 102, 106, 147, 108), ]
    #      }
    #    } else if (template == "full back") {
    #      
    #      if (nrow(data1) > 148) {
    #        
    #        data1$no <- 1:nrow(data1)
    #        data2$no <- 1:nrow(data2)
    #        data1 <- data1[c(3, 9, 10, 13, 114, 46, 47, 125, 43, 44, 147, 96, 107, 102), ]
    #        data2 <- data2[c(3, 9, 10, 13, 114, 46, 47, 125, 43, 44, 147, 96, 107, 102), ]
    #      } else {
    #        
    #        data1$no <- 1:nrow(data1)
    #        data2$no <- 1:nrow(data2)
    #        data1 <- data1[c(3, 9, 10, 13, 113, 45, 46, 124, 42, 43, 146, 95, 106, 101), ]
    #        data2 <- data2[c(3, 9, 10, 13, 114, 46, 47, 125, 43, 44, 147, 96, 107, 102), ]
    #      }
    #    } else if (template == "winger") {
    #      
    #      if (nrow(data1) > 148) {
    #        
    #        data1$no <- 1:nrow(data1)
    #        data2$no <- 1:nrow(data2)
    #        data1 <- data1[c(3, 22, 24, 42, 143, 45, 119, 47, 124, 133, 107, 146, 101, 102), ]
    #        data2 <- data2[c(3, 22, 24, 42, 143, 45, 119, 47, 124, 133, 107, 146, 101, 102), ]
    #      } else {
    #        
    #        data1$no <- 1:nrow(data1)
    #        data2$no <- 1:nrow(data2)
    #        data1 <- data1[c(3, 21, 23, 41, 142, 44, 118, 46, 123, 132, 106, 145, 100, 101), ]
    #        data2 <- data2[c(3, 22, 24, 42, 143, 45, 119, 47, 124, 133, 107, 146, 101, 102), ]
    #      }
    #    } else if (template == "goalkeeper") {
    #      
    #     if (nrow(data1) > 36) {
    #        
    #        data1$no <- 1:nrow(data1)
    #        data2$no <- 1:nrow(data2)
    #        data1 <- data1[c(4, 20, 22, 24, 26, 29, 35, 36, 37), ]
    #        data2 <- data2[c(4, 20, 22, 24, 26, 29, 35, 36, 37), ]
    #      } else {
    #        
    #        data1$no <- 1:nrow(data1)
    #        data2$no <- 1:nrow(data2)
    #        data1 <- data1[c(4, 19, 21, 23, 25, 28, 34, 35, 36), ]
    #        data2 <- data2[c(4, 20, 22, 24, 26, 29, 35, 36, 37), ]
    #      }
    #    } else if (template == "custom") {
    #      data1
    #      data2
    #    }
    
    data2 <- data2 %>%
      rename(player = Player,
             per90 = Per90,
             percentile = Percentile)
    
    player_name1 <- unique(data1$Player)
    player_name2 <- unique(data2$player)
    min1 <- unique(data1$BasedOnMinutes)
    min2 <- unique(data2$BasedOnMinutes)
    sub <- unique(data1$Versus)
    lg1 <- unique(data1$scouting_period)
    lg2 <- unique(data2$scouting_period)
    title <- paste(player_name1, "|", lg1, "|", min1, "minutes")
    subtitle <- paste(player_name2, "|", lg2, "|", min2, "minutes")
    caption <- paste("Compared to", sub, ".\nData from Stats Perform via FBref. Inspired by @FootballSlices. Created using ggshakeR.")
    
    x <- data1$Statistic
    
    data1 <- data1 %>%
      arrange(desc(stat), desc(Percentile)) %>%
      mutate(Statistic = factor(Statistic, levels = Statistic))
    
    data2 <- data2 %>%
      arrange(desc(stat), desc(percentile)) %>%
      mutate(Statistic = factor(Statistic, levels = Statistic))
    
    ggplot(data1, aes(x = Statistic, y = Percentile)) +
      geom_bar(aes(y = 100), fill = fill_b, stat = "identity", width = 1, color = gridline,
               alpha = 0.5, show.legend = FALSE) +
      geom_bar(data = data1, aes(y = Percentile, fill = color_compare), color = color_compare, stat = "identity", width = 1, alpha = 1) +
      scale_fill_manual(values = color_compare) +
      geom_bar(data = data2, aes(y = percentile, fill = NA), stat = "identity", width = 1, alpha = 0, color = colorLine, size = 3) +
      coord_polar(clip = "off") +
      geom_hline(yintercept = 25, color = colorLine, linetype = "dashed", alpha = 0.7) +
      geom_hline(yintercept = 50, color = colorLine, linetype = "dashed", alpha = 0.7) +
      geom_hline(yintercept = 75, color = colorLine, linetype = "dashed", alpha = 0.7) +
      scale_y_continuous(limits = c(-20, 100)) +
      labs(caption = caption,
           title = title,
           subtitle = subtitle) +
      theme_minimal() +
      theme(plot.background = element_rect(fill = fill_b, color = color_b),
            panel.background = element_rect(fill = fill_b, color = color_b),
            legend.position = "none",
            axis.title.y = element_blank(),
            axis.title.x = element_blank(),
            axis.text.y = element_blank(),
            axis.text.x = element_text(size = 12, color = colorText),
            text = element_text(color = colorText, size = 20),
            plot.title = element_text(hjust = 0.5, size = 26, color = color_compare, face = "bold"),
            plot.subtitle = element_text(hjust = 0.5, size = 26, color = colorLine, face = "bold"),
            plot.caption = element_text(hjust = 0.5, size = 15, color = colorText),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank()) +
      scale_x_discrete(labels = function(x) text_wrap(x = x))
  }
}